In [ ]:
# !pip install openai
# !pip install python-dotenv

In [1]:
from dotenv import load_dotenv
import os
from openai import OpenAI

In [2]:
load_dotenv()

True

In [6]:
from groq import Groq

client = Groq()
completion = client.chat.completions.create(
    model="openai/gpt-oss-20b",
    messages=[
      {
        "role": "user",
        "content": "Who won the match between India and South Africa on Dec 3rd 2025"
      }
    ],
    temperature=1,
    max_completion_tokens=8192,
    top_p=1,
    reasoning_effort="medium",
    stream=True,
    stop=None
)

for chunk in completion:
    print(chunk.choices[0].delta.content or "", end="")


I’m sorry, but I don’t have that information. My training only goes up until mid‑2024, and I can’t provide details on events that happened after that date. If you’d like to know about previous India–South Africa matches or general cricket stats, I’d be happy to help with that!

In [7]:
#client = OpenAI()
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

In [13]:
response = client.responses.create(
    model="gpt-5.1",
    input="Write a short bedtime story about a unicorn."
)
print(response.output_text)


In a quiet valley where the stars shone brighter than anywhere else, there lived a small unicorn named Liora. Her coat was soft and silver, and her horn glowed with the palest blue light, like moonlight on snow.

Every unicorn in the valley had a special gift. Some could make flowers bloom with a stomp of their hooves, others could sing songs that made rainbows appear. Liora, however, did not yet know what her gift was.

Each night, she watched the older unicorns practice their magic. She tried to help: she stomped her hooves over the grass, but no flowers came. She sang softly to the sky, but no rainbows formed. She dipped her horn into the river, hoping it would turn to sparkling sugar, but the water stayed water.

One evening, feeling a little sad, Liora wandered away from the valley and climbed a small hill. Above her, the sky was dark and deep, sprinkled with thousands of tiny stars. The moon hung low and round, like a silver lantern.

“I wish I knew what I’m meant to do,” she whi

In [ ]:
response = client.responses.create(
    model="gpt-5.1",
    input="Write a short bedtime story about a unicorn.",
    temperature=0.7, # Creativity Control : 0.0 (more focused) to 1.0 (more creative)
    top_p=1.0, # Nucleus Sampling: 0.0 to 1.0: 1.0 means consider all tokens : 0.5 means consider only top 50% probable tokens
    max_output_tokens=200,
    frequency_penalty=0.3, # Discourage Repetition: 0.0 to 2.0
    presence_penalty=0.0 # Encourage New Topics: 0.0 to 2.0
)

TypeError: Responses.create() got an unexpected keyword argument 'frequency_penalty'

In [53]:
## Use Chat Completion not Chat Response
from openai import OpenAI
client = OpenAI()

response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[{"role": "user", "content": "Write a bedtime story about a unicorn"}],
    temperature=0.7,
    top_p=1.0,
    frequency_penalty=0.5,
    presence_penalty=0.3
)




In [52]:
print(response.choices[0].message.content)

**The Moonlit Journey of Luna the Unicorn**

Once upon a time in a mystical land called Celestia, where the skies shimmered with rainbow colors and the trees whispered secrets to one another, there lived a gentle unicorn named Luna. She had a coat as white as freshly fallen snow and a mane that sparkled like stardust. But what made Luna truly special was her magnificent horn, which glowed softly in the moonlight, casting a silver light over everything around her.

Luna loved to wander through the enchanted forest that surrounded her home. Each evening, as the sun dipped below the horizon and painted the sky with hues of orange and pink, she would set off on adventures, exploring hidden meadows filled with flowers that danced in the breeze.

One evening, while wandering deeper into the forest than she had ever been before, Luna stumbled upon an ancient oak tree with twisting roots and shimmering leaves. As she approached, she heard a soft sobbing sound coming from beneath its branches. 

## Test Transcripts

In [1]:
from openai import OpenAI
client = OpenAI()

audio_file = open("speech_test_regenerated.mp3", "rb")

transcript = client.audio.transcriptions.create(
    model="gpt-4o-transcribe",
    file=audio_file
)

print(transcript)


Transcription(text='Hi there, this is a regenerated audio sample for testing the OpenAI transcription API. This audio is short, clear, and suitable for your transcription experiments.', logprobs=None, usage=UsageTokens(input_tokens=88, output_tokens=33, total_tokens=121, type='tokens', input_token_details=UsageTokensInputTokenDetails(audio_tokens=88, text_tokens=0)))


# GROQ

### Set the key in .env
### load env

In [4]:
#!pip install groq

In [7]:
from dotenv import load_dotenv
import os
load_dotenv()

True

In [60]:
from groq import Groq

client = Groq()
completion = client.chat.completions.create(
    model="llama-3.1-8b-instant",
    messages=[
      {
        "role": "user",
        "content": ""
      }
    ],
    temperature=1,
    max_completion_tokens=1024,
    top_p=1,
    stream=True,
    stop=None
)

for chunk in completion:
    print(chunk.choices[0].delta.content or "", end="")


Is there something I can help you with?

In [9]:
client = Groq()
completion = client.chat.completions.create(
    model="llama-3.1-8b-instant",
    messages=[
      {
        "role": "user",
        "content": "Share the latest news about AI research from the past month."
      }
    ],
    temperature=1,
    max_completion_tokens=1024,
    top_p=1,
    stream=True,
    stop=None
)

for chunk in completion:
    print(chunk.choices[0].delta.content or "", end="")


Since my knowledge cutoff is December 2023, I'll provide information on AI research news from the period leading up to that date. However, please note that some of the information might have changed or been updated since then.

Here are some of the latest developments in AI research from the past few months (up to December 2023):

1. **Breakthroughs in Large Language Models:** Researchers have made significant progress in developing more efficient and accurate large language models (LLMs). For example, Meta AI's LLaMA model has shown excellent performance in various NLP tasks, and the researchers have also made improvements to the model architecture.

2. **Advances in Explainable AI (XAI):** Researchers at Google have made strides in developing more interpretable and explainable AI models. They have created a new tool, Explainable Neural Networks (ENN), which can provide human-interpretable explanations for AI decisions.

3. **AI for Cybersecurity:** Researchers at the University of Ca

In [11]:
!pip install ollama

# deepseek-r1:1.5b Using OLlama

In [15]:
from ollama import chat
from ollama import ChatResponse

response: ChatResponse = chat(model='deepseek-r1:1.5b', messages=[
  {
    'role': 'user',
    'content': 'Why is the sky blue?',
  },
])
#print(response['message']['content'])
# or access fields directly from the response object
print(response.message.content)

The color of the sky, known as coloration or chromacy, is primarily determined by the amount of sunlight reflected on the Earth's surface. Different parts of the sky have varying levels of this reflection because light reflecting off water particles in the atmosphere, such as clouds and mist, can pass through more air molecules than light from objects like buildings or cars that reflects off dust or other particles. This difference in reflection leads to slight variations in the hues we perceive when looking up at the sky. Additionally, atmospheric conditions, including temperature, humidity, and the presence of trace gases like carbon dioxide, influence how the sky appears.


# Using Langchain

In [18]:
#!pip install langchain-openai

In [19]:
from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate

llm = ChatOpenAI(model="gpt-4o-mini")

prompt = ChatPromptTemplate.from_messages([
    ("system", "You are a helpful assistant."),
    ("human", "{question}")
])

chain = prompt | llm
print(chain.invoke({"question": "What is Agentic AI?"}))


content='Agentic AI refers to artificial intelligence systems that possess the ability to take autonomous actions, make decisions, and interact with the environment to achieve specific goals or tasks. Unlike traditional AI, which often operates under human direction and focuses on tasks like data analysis or pattern recognition, agentic AI can operate independently and adaptively.\n\nKey characteristics of agentic AI may include:\n\n1. **Autonomy**: The AI can perform tasks and make decisions without human intervention.\n2. **Goal-oriented behavior**: It is designed to achieve specific objectives or goals, which could involve planning and decision-making.\n3. **Adaptability**: Agentic AI can learn from its experiences and change its behavior based on new information or feedback from its environment.\n4. **Interactivity**: These systems can often interact with complex environments or with humans and other agents to achieve their objectives.\n\nExamples of agentic AI can range from advan

In [22]:
#pip install langchain-groq

In [9]:
from langchain_groq import ChatGroq
llm = ChatGroq(model="llama-3.1-8b-instant")

response = llm.invoke("Explain transformers in simple words.")
print(response)


content='**What are Transformers?**\n\nTransformers are a type of neural network model used in machine learning and artificial intelligence. They\'re called "transformers" because they transform the way we process and understand language (and other data).\n\n**How do Transformers Work?**\n\nImagine you\'re having a conversation with a friend. You start with a sentence, and your friend responds with another sentence. But, how do they know what word to use next? They use context, right? They consider what you said before and try to make sense of it.\n\nTransformers work in a similar way. They look at a piece of text (like a sentence or a paragraph) and break it down into smaller chunks called "tokens" (like individual words). Then, they use these tokens to predict what comes next, just like your friend does.\n\nHere\'s a simplified step-by-step explanation:\n\n1. **Tokenization**: Break down the text into individual words or tokens.\n2. **Embedding**: Convert each token into a numerical 

## System Prompt

In [25]:
from langchain_core.prompts import PromptTemplate

template = PromptTemplate(
    input_variables=["topic"],
    template="Explain {topic} in 3 bullet points."
)

print(template.format(topic="embeddings"))


Explain embeddings in 3 bullet points.


In [26]:
from langchain_core.output_parsers import JsonOutputParser
from langchain_openai import ChatOpenAI

parser = JsonOutputParser()

prompt = """
Extract company name and revenue from the text below.
Return output ONLY in JSON format with keys company, revenue.

Text:
Apple generated $383B revenue in 2023.
"""

model = ChatOpenAI()
print(parser.parse(model.invoke(prompt).content))


{'company': 'Apple', 'revenue': '$383B'}


# Embeddings

In [27]:
from openai import OpenAI

client = OpenAI()

# Example text list
texts = [
    "Retrieval-Augmented Generation (RAG) improves accuracy.",
    "Embeddings convert text into high-dimensional vectors."
]

response = client.embeddings.create(
    model="text-embedding-3-small",
    input=texts
)

# Print embeddings
for i, emb in enumerate(response.data):
    print(f"Text {i+1} embedding length:", len(emb.embedding))
    print(emb.embedding[:10], "...")  # show first 10 numbers


Text 1 embedding length: 1536
[0.015356769785284996, 0.03452569618821144, 0.03314683213829994, 0.01639767736196518, -0.021115558221936226, -0.0155595438554883, 0.013437173329293728, 0.012734223157167435, -0.018749859184026718, 0.019087815657258034] ...
Text 2 embedding length: 1536
[-0.027849717065691948, 0.003486940171569586, 0.04356098920106888, 0.0014729317044839263, 0.014623391442000866, -0.01214989647269249, 0.015505146235227585, -0.00042477401439100504, 0.002828486729413271, 0.023143205791711807] ...


In [54]:
response = client.embeddings.create(
    model="text-embedding-3-small",
    input=texts,
    dimensions=512
)

# Print embeddings
for i, emb in enumerate(response.data):
    print(f"Text {i+1} embedding length:", len(emb.embedding))
    print(emb.embedding[:10], "...")  # show first 10 numbers


Text 1 embedding length: 512
[-0.006820265203714371, 0.01068376749753952, -0.022077158093452454, 0.030908022075891495, -0.06736687570810318, -0.03038763254880905, 0.06591609120368958, 0.033778052777051926, -0.00962721835821867, -0.008815093897283077] ...
Text 2 embedding length: 512
[-0.03966500610113144, -0.028670484200119972, 0.03462023660540581, -0.05522293224930763, 0.05333595722913742, -0.043862566351890564, -0.04640420898795128, -0.005848662927746773, -0.045826561748981476, 0.06434973329305649] ...
Text 3 embedding length: 512
[-0.005087872035801411, 0.039913903921842575, 0.016241736710071564, -0.00626326072961092, -0.004804299212992191, -0.030938206240534782, 0.03853302821516991, -0.001427111099474132, -0.020976169034838676, 0.027683284133672714] ...


#### Size
    1. text-embedding-3-large : Default vector size: 3072 dimensions [Large]
    2. text-embedding-3-small : 1536 dimensions
    

In [30]:
#!pip install langchain-community

In [31]:
from langchain_openai import OpenAIEmbeddings
from langchain_community.vectorstores import FAISS




In [32]:
# Step 1. Embedding model
embedding_model = OpenAIEmbeddings(model="text-embedding-3-small")

# Step 2. Sample documents
docs = [
    "India won its Test match by 200 runs.",
    "RAG improves retrieval accuracy for enterprise systems.",
    "Transfer learning speeds up deep learning training."
]

# Step 3. Create vector store
faiss_index = FAISS.from_texts(docs, embedding_model)



Top results:
- Transfer learning speeds up deep learning training.
- RAG improves retrieval accuracy for enterprise systems.


In [33]:
# Step 4. Search
query = "How do I improve AI accuracy?"
results = faiss_index.similarity_search(query, k=2)

print("Top results:")
for r in results:
    print("-", r.page_content)

Top results:
- Transfer learning speeds up deep learning training.
- RAG improves retrieval accuracy for enterprise systems.


In [34]:
db = FAISS.from_texts(docs, embedding_model)

In [35]:
# Save it
db.save_local("faiss_index")

In [36]:
# Load existing FAISS index
db = FAISS.load_local("faiss_index", embedding_model, allow_dangerous_deserialization=True)


In [37]:
docs_new = [
    "AI can automate repetitive tasks.",
    "Machine learning models are improving healthcare outcomes.",
    "The stock market crashed yesterday due to global tension.",
    "Cats are very playful animals.",
    "Python is widely used in data science.",
    "Electric vehicles are becoming more popular.",
    "The weather today is sunny with mild winds.",
    "Football is the most popular sport in the world.",
    "Quantum computing is the future of cryptography.",
    "Deep learning helps in image recognition.",
    "The restaurant serves Italian and Mexican food.",
    "Traveling to Japan is on my wishlist.",
    "Reinforcement learning trains agents via rewards.",
    "Natural Language Processing enables chatbots.",
    "The new iPhone was launched with major upgrades.",
    "Meditation improves mental health.",
    "Blockchain technology ensures secure transactions.",
    "Yoga enhances flexibility and strength.",
    "Cloud computing allows scalable applications.",
    "Climate change impacts global ecosystems."
]

In [38]:
db.add_texts(docs_new)

['fedbf86f-e0f3-4aa6-82a8-9438d174773f',
 '64fba312-22a9-440d-921e-1c029df9593d',
 '892017d6-430c-4215-abfc-e7dd386749f9',
 'bbb04f95-0d53-4a0c-b4b0-dd6813cac7c9',
 'dedf16ef-62e1-46af-b8b4-b0a663ca0d31',
 'adc02ff3-8444-4583-9306-ab0a636b2f56',
 '55cfc178-a97c-41a2-9eff-d259ba1d4613',
 '5e6cb105-5166-4390-8617-86f0ffcfed28',
 '59cb40b5-4b3e-4d25-a79d-b95128d3afed',
 '5a585964-b038-4b3b-8a7d-69775f19282d',
 '511f5dd6-d287-42c1-a5ab-facff615d06f',
 '31b7c22b-d3ee-486f-aa7e-dc7f1710ef68',
 'ecf6ba3d-a2e2-478f-a817-a4dc98ac4e27',
 '23b4fdaf-88b3-4977-bc0e-34d03675b16a',
 '2d26d37c-bcb1-40cc-9523-cd000ecea548',
 'c3158f71-c5aa-4f3f-99e8-f4661c5e8e8d',
 '4c805f06-82ca-4078-ae92-e320e5edbaff',
 'b0c43edd-4a34-4866-9bea-5e0594b92177',
 'abe38cf7-043e-470a-84c2-ed29fa58ddf6',
 '0a0be554-5325-468e-a1c5-0f40917ec87e']

In [39]:
# Save updated index
db.save_local("faiss_index")

In [41]:
query = "How can AI help in healthcare?"
results = db.similarity_search(query, k=3)
print("\nTop search results:")
for r in results:
    print("-", r.page_content)


Top search results:
- Machine learning models are improving healthcare outcomes.
- AI can automate repetitive tasks.
- Deep learning helps in image recognition.


In [42]:
query = "How do vector databases work?"

results = db.similarity_search_with_score(query, k=3)

for doc, score in results:
    print("Text:", doc.page_content)
    print("Score:", score)
    print("---")


Text: RAG improves retrieval accuracy for enterprise systems.
Score: 1.3513801
---
Text: Cloud computing allows scalable applications.
Score: 1.5137949
---
Text: Transfer learning speeds up deep learning training.
Score: 1.5689678
---


In [43]:
print("FAISS index type:", db.index.metric_type)


FAISS index type: 1


#### distance
1 -- Inner Product (higher is better)

0 -- L2 Distance (Smaller is better)

In [45]:
from langchain_openai import OpenAIEmbeddings
from langchain_community.vectorstores import FAISS
from langchain_community.vectorstores.utils import DistanceStrategy

emb = OpenAIEmbeddings(model="text-embedding-3-small")

texts = [
    "India won the cricket match.",
    "FAISS is used for similarity search.",
    "Embeddings turn text into vectors."
]

db = FAISS.from_texts(
    texts,
    emb,
    distance_strategy=DistanceStrategy.COSINE
)

print("Metric type:", db.index.metric_type)


Metric type: 1


#### Different Distance
    1. DistanceStrategy.EUCLIDEAN (L2)
        - L2
        - Lower = closer
        - Useful for numeric vectors, clustering
    2. DistanceStrategy.COSINE
        - Equivalent to cosine similarity
        - Higher = closer
        - Best for RAG, semantic search
    3. DistanceStrategy.MAX_INNER_PRODUCT
        - Higher = similar
        - Good for ranking tasks
        - Recommendation Systems (weighted embeddings), Product Ranking on E-commerce Platforms 


In [ ]:
from langchain_community.vectorstores.utils import DistanceStrategy

# Cosine similarity
db_cos = FAISS.from_texts(texts, emb, distance_strategy=DistanceStrategy.COSINE)

# Euclidean distance
db_l2 = FAISS.from_texts(texts, emb, distance_strategy=DistanceStrategy.EUCLIDEAN)

# Max inner-product
db_ip = FAISS.from_texts(texts, emb, distance_strategy=DistanceStrategy.MAX_INNER_PRODUCT)


## Effective Prompt
#### Examples

In [56]:
from openai import OpenAI

# Initialize client
client = OpenAI()

# Mock RAG retrieved content
retrieved_regulation_chunks = """
Section 4.1: Transactions above $10,000 must be reported within 24 hours.
Section 5.3: All foreign transactions require dual authorization.
"""

# User question
audit_team_question = "Do we need to report domestic transactions above $15,000?"

# Construct Better Prompt
prompt = f"""
You are a Financial Compliance Assistant for internal audit teams.
Use ONLY the regulatory text provided in the CONTEXT section.
For the given QUESTION, follow these rules:

1. Provide a concise answer in 2–3 sentences.
2. Cite the specific regulation clause numbers or section titles you used.
3. If the regulation does not explicitly mention the requested information, state:
   "The regulation does not explicitly cover this point."
4. Do NOT guess or infer new rules. Do NOT provide advice beyond the context.

CONTEXT:
{retrieved_regulation_chunks}

QUESTION:
{audit_team_question}
"""

# Invoke LLM
response = client.responses.create(
    model="gpt-4o-mini",
    input=prompt,
    max_output_tokens=250
)

# Output
print(response.output_text)


The regulation does not explicitly cover this point. Section 4.1 specifies reporting requirements for transactions above $10,000 but does not differentiate between domestic and foreign transactions.


In [57]:
from openai import OpenAI

client = OpenAI()

# Mock HR policy chunk (normally retrieved via RAG)
retrieved_hr_policy_chunks = """
Section 2.3: Employees can take up to 12 paid sick leaves per calendar year.
Section 4.1: Work-from-home is allowed for up to 8 days per month with manager approval.
Section 5.4: Overtime pay applies only when weekly work exceeds 40 hours.
"""

employee_question = "Am I allowed to work from home for 10 days in a month?"

prompt = f"""
You are an Internal HR Policy Assistant.
Use ONLY the HR policy text provided in the CONTEXT section.
For the given QUESTION, follow these rules:

1. Provide a clear answer in 2–4 sentences.
2. Cite the exact HR policy section numbers used.
3. If the policy does not explicitly cover the question, say:
   "The policy does not explicitly cover this issue."
4. Do NOT invent new HR rules or benefits.

CONTEXT:
{retrieved_hr_policy_chunks}

QUESTION:
{employee_question}
"""

response = client.responses.create(
    model="gpt-4o-mini",
    input=prompt,
    max_output_tokens=200
)

print(response.output_text)


No, you are not allowed to work from home for 10 days in a month, as the policy permits work-from-home for up to 8 days per month with manager approval (Section 4.1).


In [58]:
from openai import OpenAI

client = OpenAI()

# Mock Security policy chunk (normally from vector DB)
retrieved_security_policy_chunks = """
Control ID SEC-12: All company laptops must use full-disk encryption.
Control ID SEC-18: Passwords must be changed every 90 days.
Control ID SEC-27: Multi-factor authentication (MFA) is mandatory for VPN access.
"""

it_security_question = "Is multi-factor authentication required for accessing company email?"

prompt = f"""
You are an Internal Information Security Policy Assistant.
Use ONLY the security policy text provided in the CONTEXT section.
For the given QUESTION, follow these rules:

1. Provide a precise answer in 2–4 sentences.
2. Cite the exact control IDs you used.
3. If the policy does not directly address the question, state:
   "The policy does not explicitly address this issue."
4. Do NOT invent new security requirements or controls.
5. Maintain compliance-ready, factual language.

CONTEXT:
{retrieved_security_policy_chunks}

QUESTION:
{it_security_question}
"""

response = client.responses.create(
    model="gpt-4o-mini",
    input=prompt,
    max_output_tokens=200
)

print(response.output_text)


The policy does not explicitly address this issue.


## Types
    1. Zero Shot
    2. Few Shot
    3. Chain of Thoughts
    4. Schema guided prompting

In [1]:
from openai import OpenAI
client = OpenAI()

prompt = "Classify the sentiment of the message: 'The delivery was late and the product was damaged.'"

response = client.responses.create(
    model="gpt-4o-mini",
    input=prompt
)

print(response.output_text)


The sentiment of the message is negative.


In [2]:
from openai import OpenAI
client = OpenAI()

prompt = """
Classify the IT support issue into one of: [Network, Hardware, Software]

Example:
Input: "WiFi keeps dropping."
Output: Network

Example:
Input: "My laptop screen is cracked."
Output: Hardware

Now classify:
Input: "Outlook is not syncing emails."
Output:
"""

response = client.responses.create(
    model="gpt-4o-mini",
    input=prompt
)

print(response.output_text)


Software


In [8]:
from openai import OpenAI
client = OpenAI()

COTprompt = """
Think step-by-step and solve this:

A company made ₹80 lakh revenue last year.
This year revenue increased to ₹120 lakh.
What is the percentage growth?
"""

response = client.responses.create(
    model="gpt-4o-mini",
    input=COTprompt
)

print(response.output_text)


To calculate the percentage growth in revenue, you can use the following formula:

\[
\text{Percentage Growth} = \left( \frac{\text{New Revenue} - \text{Old Revenue}}{\text{Old Revenue}} \right) \times 100
\]

1. **Identify the old and new revenue:**
   - Old Revenue = ₹80 lakh
   - New Revenue = ₹120 lakh

2. **Calculate the difference:**
   - Difference = New Revenue - Old Revenue
   - Difference = ₹120 lakh - ₹80 lakh = ₹40 lakh

3. **Plug the values into the percentage growth formula:**
   \[
   \text{Percentage Growth} = \left( \frac{₹40 \text{ lakh}}{₹80 \text{ lakh}} \right) \times 100
   \]

4. **Calculate:**
   \[
   \text{Percentage Growth} = 0.5 \times 100 = 50\%
   \]

Therefore, the percentage growth in revenue is **50%**.


#### How to fix the issue

In [9]:
from langchain_core.output_parsers import JsonOutputParser
parser = JsonOutputParser()
JSonprompt = """
Compute revenue growth and return ONLY JSON:
{
  "old": 0,
  "new": 0,
  "growth_pct": 0
}

Old revenue: 80
New revenue: 120

Do not include text or explanation.
"""

response = client.responses.create(
    model="gpt-4o-mini",
    input=JSonprompt
)



In [10]:
result = parser.parse(response.output_text)
print(result)

{'old': 80, 'new': 120, 'growth_pct': 50}


## TOOLS and AGENTS

In [13]:
!pip show langchain-community

Name: langchain-community
Version: 0.0.20
Summary: Community contributed LangChain integrations.
Home-page: https://github.com/langchain-ai/langchain
Author: 
Author-email: 
License: MIT
Location: /opt/anaconda3/envs/AAI/lib/python3.12/site-packages
Requires: aiohttp, dataclasses-json, langchain-core, langsmith, numpy, PyYAML, requests, SQLAlchemy, tenacity
Required-by: langchain


In [25]:
!pip install wikipedia
!pip install -U langchain langchain-core langchain-community langchain-openai langgraph


  Using cached langchain_core-1.1.0-py3-none-any.whl.metadata (3.6 kB)
  Using cached langchain_community-0.4.1-py3-none-any.whl.metadata (3.0 kB)
  Using cached langsmith-0.4.49-py3-none-any.whl.metadata (14 kB)
Using cached langchain_core-1.1.0-py3-none-any.whl (473 kB)
Using cached langsmith-0.4.49-py3-none-any.whl (410 kB)
Using cached langchain_community-0.4.1-py3-none-any.whl (2.5 MB)
  Attempting uninstall: langsmith
    Found existing installation: langsmith 0.0.87
    Uninstalling langsmith-0.0.87:
      Successfully uninstalled langsmith-0.0.87
  Attempting uninstall: langchain-core━━━━━━━━━━━━━━━━━━━━━━━━━━━━  2/10 [langsmith]
    Found existing installation: langchain-core 0.1.23━━━━━━━━  2/10 [langsmith]
    Uninstalling langchain-core-0.1.23:━━━━━━━━━━━━━━━━━━━━━━━  2/10 [langsmith]
      Successfully uninstalled langchain-core-0.1.23━━━━━━━━━━  2/10 [langsmith]
  Attempting uninstall: langchain-community━━━━━━━━━━━━━━━  6/10 [langgraph-prebuilt]
    Found existing instal

In [2]:
pip install -U ddgs

  Using cached socksio-1.0.0-py3-none-any.whl.metadata (6.1 kB)
  Using cached hyperframe-6.1.0-py3-none-any.whl.metadata (4.3 kB)
  Using cached hpack-4.1.0-py3-none-any.whl.metadata (4.6 kB)
Using cached hpack-4.1.0-py3-none-any.whl (34 kB)
Using cached hyperframe-6.1.0-py3-none-any.whl (13 kB)
Using cached socksio-1.0.0-py3-none-any.whl (12 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 5.1 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7/7 [ddgs]6/7 [ddgs]
Note: you may need to restart the kernel to use updated packages.


In [18]:
pip install youtube-search

Note: you may need to restart the kernel to use updated packages.


In [38]:
# Install required packages:
# pip install langchain langchain-community langchain-openai

from langchain_openai import ChatOpenAI

# Predefined community tools
from langchain_community.tools import DuckDuckGoSearchRun
#from langchain_community.tools import WikipediaQueryRun
#from langchain_community.tools import ArxivQueryRun
from langchain_community.tools.tavily_search import TavilySearchResults
from langchain_community.tools import YouTubeSearchTool

from langchain_community.utilities import WikipediaAPIWrapper, ArxivAPIWrapper


# Intitialize Tools
duckduckgo = DuckDuckGoSearchRun()
tavily = TavilySearchResults(max_results=3,)
youtube = YouTubeSearchTool()
tools = [duckduckgo, tavily, youtube]


In [39]:
# Bind Tools to LLM
llm = ChatOpenAI(model="gpt-4o-mini",  temperature=0).bind_tools(tools)

In [40]:
llm

RunnableBinding(bound=ChatOpenAI(profile={'max_input_tokens': 128000, 'max_output_tokens': 16384, 'image_inputs': True, 'audio_inputs': False, 'video_inputs': False, 'image_outputs': False, 'audio_outputs': False, 'video_outputs': False, 'reasoning_output': False, 'tool_calling': True, 'structured_output': True, 'image_url_inputs': True, 'pdf_inputs': True, 'pdf_tool_message': True, 'image_tool_message': True, 'tool_choice': True}, client=<openai.resources.chat.completions.completions.Completions object at 0x17be4f980>, async_client=<openai.resources.chat.completions.completions.AsyncCompletions object at 0x17c3d8260>, root_client=<openai.OpenAI object at 0x17bfde1b0>, root_async_client=<openai.AsyncOpenAI object at 0x17c3d85c0>, model_name='gpt-4o-mini', temperature=0.0, model_kwargs={}, openai_api_key=SecretStr('**********'), stream_usage=True), kwargs={'tools': [{'type': 'function', 'function': {'name': 'duckduckgo_search', 'description': 'A wrapper around DuckDuckGo Search. Useful 

In [41]:
print("\n=== 1. DuckDuckGo Search ===")
print(llm.invoke("Search for the latest AI news in India."))


=== 1. DuckDuckGo Search ===
content='' additional_kwargs={'refusal': None} response_metadata={'token_usage': {'completion_tokens': 22, 'prompt_tokens': 205, 'total_tokens': 227, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_provider': 'openai', 'model_name': 'gpt-4o-mini-2024-07-18', 'system_fingerprint': 'fp_b547601dbd', 'id': 'chatcmpl-ChVxOKWcMUZPBMqXsmAraCIW6oqXh', 'service_tier': 'default', 'finish_reason': 'tool_calls', 'logprobs': None} id='lc_run--d9a9fe28-134e-4a6d-872b-db02a5676904-0' tool_calls=[{'name': 'tavily_search_results_json', 'args': {'query': 'latest AI news in India'}, 'id': 'call_smPJd23vWwwOrVixinVSReDF', 'type': 'tool_call'}] usage_metadata={'input_tokens': 205, 'output_tokens': 22, 'total_tokens': 227, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 're

In [6]:
print("\n=== 2. Tavily Web Search (factual search) ===")
print(llm.invoke("Find the top 3 facts about renewable energy market 2025."))


=== 2. Tavily Web Search (factual search) ===
content='' additional_kwargs={'refusal': None} response_metadata={'token_usage': {'completion_tokens': 25, 'prompt_tokens': 210, 'total_tokens': 235, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_provider': 'openai', 'model_name': 'gpt-4o-mini-2024-07-18', 'system_fingerprint': 'fp_b547601dbd', 'id': 'chatcmpl-ChT3Na8XawnSabmIC0CzK7WrC8iU5', 'service_tier': 'default', 'finish_reason': 'tool_calls', 'logprobs': None} id='lc_run--01fe5bdc-9f18-41a3-a719-cf3bc186bb44-0' tool_calls=[{'name': 'tavily_search_results_json', 'args': {'query': 'renewable energy market 2025 facts'}, 'id': 'call_ncIFwBR09VdhiVbyQHb4c4jp', 'type': 'tool_call'}] usage_metadata={'input_tokens': 210, 'output_tokens': 25, 'total_tokens': 235, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_toke

In [7]:
print("\n=== 3. YouTube Search ===")
print(llm.invoke("Find useful YouTube videos for beginners in Python programming."))


=== 3. YouTube Search ===
content='' additional_kwargs={'refusal': None} response_metadata={'token_usage': {'completion_tokens': 17, 'prompt_tokens': 207, 'total_tokens': 224, 'completion_tokens_details': {'accepted_prediction_tokens': 0, 'audio_tokens': 0, 'reasoning_tokens': 0, 'rejected_prediction_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': 0, 'cached_tokens': 0}}, 'model_provider': 'openai', 'model_name': 'gpt-4o-mini-2024-07-18', 'system_fingerprint': 'fp_560af6e559', 'id': 'chatcmpl-ChT3QgOhkLw4EsuyUbyluzQGmr6mt', 'service_tier': 'default', 'finish_reason': 'tool_calls', 'logprobs': None} id='lc_run--4d63f2db-2db5-446c-8eb0-d6e37a584ce4-0' tool_calls=[{'name': 'youtube_search', 'args': {'query': 'Python programming for beginners'}, 'id': 'call_C04Qg3uCDGeAsqE6EcUyCrPK', 'type': 'tool_call'}] usage_metadata={'input_tokens': 207, 'output_tokens': 17, 'total_tokens': 224, 'input_token_details': {'audio': 0, 'cache_read': 0}, 'output_token_details': {'audio': 0, 'reasonin

## For Every human Message LLM will identify which tool to be used and then invoke the tool

In [42]:
from langchain_core.messages import HumanMessage
tools = [duckduckgo, tavily, youtube]

In [45]:
def ask_ai(query):
    print(f"\n=== USER QUERY: {query} ===")

    result = llm.invoke([HumanMessage(content=query)])

    # FIX: Read tool calls from result.tool_calls
    tool_calls = result.tool_calls or []
    print(tool_calls)

    if tool_calls:
        call = tool_calls[0]
        tool_name = call["name"]
        tool_args = call["args"]

        print(f"\n🔧 LLM SELECTED TOOL → {tool_name}")
        print(f"🔢 Tool Arguments → {tool_args}")

        # ---- Run the actual tool ----
        if "duck" in tool_name.lower():
            tool_output = duckduckgo.run(tool_args["query"])

        elif "tavily" in tool_name.lower():
            tool_output = tavily.run(tool_args)

        elif "youtube" in tool_name.lower():
            tool_output = youtube.run(tool_args["query"])

        else:
            tool_output = f"❓ Unknown tool selected → {tool_name}"

        print("\n🛠 TOOL OUTPUT:")
        print(tool_output)

        # Final LLM polishing
        final_answer = llm.invoke(
            f"User query: {query}\nTool output: {tool_output}\nGive a clean final answer."
        )
        print("\n🧠 FINAL ANSWER:")
        print(final_answer.content)

    else:
        print("\n🤖 LLM responded WITHOUT a tool:")
        print(result.content)


In [46]:
ask_ai("Give me recent research about quantum computing.")


=== USER QUERY: Give me recent research about quantum computing. ===
[{'name': 'tavily_search_results_json', 'args': {'query': 'recent research in quantum computing 2023'}, 'id': 'call_76Lyfz5PyMZYlsqQ9lKJRQTJ', 'type': 'tool_call'}]

🔧 LLM SELECTED TOOL → tavily_search_results_json
🔢 Tool Arguments → {'query': 'recent research in quantum computing 2023'}

🛠 TOOL OUTPUT:
[{'title': 'Breakthroughs in Quantum Computing - Wevolver', 'url': 'https://www.wevolver.com/article/breakthroughs-in-quantum-computing', 'content': 'Another recent pivotal development in the quantum computing field was the shift from emphasizing processor benchmarks to focusing on practical implementations. This transition is vital for transforming quantum computing from a theoretical concept into an applied technology with real-world applications. An updated analysis for 2023 projects indicates that industries like automotive, chemicals, financial services, and life sciences could see an economic impact of up to $1.

In [51]:
ask_ai("Suggest YouTube videos to learn Python.")


=== USER QUERY: Suggest YouTube videos to learn Python. ===
[]

🤖 LLM responded WITHOUT a tool:
Here are some popular YouTube channels and specific video recommendations to help you learn Python:

### YouTube Channels

1. **Corey Schafer**
   - Known for clear explanations and practical examples.
   - **Recommended Video**: "Python Tutorial for Beginners - Full Course in 11 Hours [2020]"

2. **freeCodeCamp.org**
   - Offers comprehensive tutorials and full courses.
   - **Recommended Video**: "Python for Everybody - Full Course with Dr. Charles Severance"

3. **Programming with Mosh**
   - Focuses on practical programming skills.
   - **Recommended Video**: "Python Tutorial for Beginners - Learn Python in 1 Hour"

4. **Traversy Media**
   - Covers a wide range of programming topics, including Python.
   - **Recommended Video**: "Python Crash Course - Learn Python in 1 Hour"

5. **Sentdex**
   - Focuses on Python for data analysis, machine learning, and more.
   - **Recommended Video**

### Custom Tool

In [48]:
from langchain_openai import ChatOpenAI
from langchain.tools import tool
from langgraph.prebuilt import create_react_agent

# -------------------------
# 1. Define tools
# -------------------------
@tool
def multiply(a: int, b: int) -> int:
    """Multiply two numbers."""
    return a * b

@tool
def greet(name: str) -> str:
    """Return greeting."""
    return f"Hello {name}! Welcome."

tools = [multiply, greet]

# -------------------------
# 2. LLM
# -------------------------
llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)

# -------------------------
# 3. Build Agent (LangGraph ReAct Agent)
# -------------------------
agent = create_react_agent(llm, tools)

# -------------------------
# 4. Run agent
# -------------------------
result = agent.invoke({"messages": "Multiply 9 and 7, then greet Sahil."})

print(result["messages"][-1].content)


/var/folders/m8/lr5c6v793qbcszm54dsf6y440000gn/T/ipykernel_43602/341687680.py:28: LangGraphDeprecatedSinceV10: create_react_agent has been moved to `langchain.agents`. Please update your import to `from langchain.agents import create_agent`. Deprecated in LangGraph V1.0 to be removed in V2.0.
  agent = create_react_agent(llm, tools)


The result of multiplying 9 and 7 is 63. 

Also, hello Sahil! Welcome.


In [50]:
result = agent.invoke({"messages": "Add 10 and 7"})

print(result["messages"][-1].content)


The result of multiplying 10 and 7 is 70. 

If you meant to add 10 and 7, the sum is 17.


In [32]:
!pip install tavily-python
!pip install langchain-community


In [36]:
!pip install -U langchain langchain-community langchain-openai langgraph tavily-python
